Develop

JAKLI78 opened the pull request.

JAKLI78 committed:
add Repositores for All
add UserList and NewUser controller with views

JAKLI78 committed:
add Edit/Creat/Update user(update without Titels)

JAKLI78 committed:
using structuremap
using owin
rename floaders

JAKLI78 committed:
remove comment block

JAKLI78 committed:
make service for each repository
add async controller for load images
add async methods in user repository
add ImageService

JAKLI78 committed:
make service for each repository
add async controller for load images
add async methods in user repository
add ImageService

JAKLI78 committed:
make service for each repository
add async controller for load images
add async methods in user repository
add ImageService

JAKLI78 committed:
fix model binding from view

JAKLI78 committed:
add Search servive
add FluentValidation

JAKLI78 committed:
improve user validation

JAKLI78 committed:
add custom filters whith log function

JAKLI78 committed:
add pagination for Index and Search
true action log filter
difrent models for NewEditUser and Users
rename methods and vars
remove unused fields from company model

JAKLI78 committed:
remove second gitignore

JAKLI78 committed:
add remains global filters
one controller for crud and list
max count of displaed users in .config
photo path in .config
fix search servise
add style file
js sctipts in separate files
remove not used files

stas-i commented on MVCTask/MVCTask.Core/Interface/IImageService.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    public interface IImageService
+    {
+        Task<byte[]> GetImageAsync(string path);
```
Method name does not meet expectations. You are resizing images

stas-i commented on MVCTask/MVCTask.Core/Interface/ISearchService.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace MVCTask.Core.Interface
+{
+    public interface ISearchService
```
It could be done in user service

stas-i commented on MVCTask/MVCTask.Core/Interface/ITitelsServise.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace MVCTask.Core.Interface
+{
+    public interface ITitelsServise
```
https://translate.google.by/?hl=ru&tab=wT&google_abuse=GOOGLE_ABUSE_EXEMPTION%3DID%3D6e247fb84207631f:TM%3D1495616858:C%3Dr:IP%3D46.216.180.98-:S%3DAPGng0sO5dlyagbWb9mIjXrGtAWtISSS1g%3B+path%3D/%3B+domain%3Dgoogle.com%3B+expires%3DWed,+24-May-2017+12:07:38+GMT#en/ru/Titel

stas-i commented on MVCTask/MVCTask.Core/Interface/IUserService.cs:
```diff
@@ -0,0 +1,23 @@
+        IEnumerable<User> GetUsers();
+
+        void CreateUser(string name, string surname, string email, DateTime birthDate, int companyId,
+            string fileUrl);
```
Use model - it's hard to use lots of string parameters and hard to refactor(what if we need to add one more field to user)

stas-i commented on MVCTask/MVCTask.Core/Registers/DataRegistry.cs:
```diff
@@ -0,0 +1,20 @@
+    {
+        public DataRegistry()
+        {
+            For<IUserRepository>().Use<UserRepository>().LifecycleIs<UniquePerRequestLifecycle>();
```
Why not transient?

stas-i commented on MVCTask/MVCTask.Core/Services/ImageService.cs:
```diff
@@ -0,0 +1,34 @@
+            var img = new Bitmap(fileStream);
+            fileStream.Close();
+
+            return await ImageResizeAsync(img);
```
Do we need to await here?

stas-i commented on MVCTask/MVCTask.Core/Services/SearchService.cs:
```diff
@@ -0,0 +1,55 @@
+            var result = new List<User>();
+            var companys = _companyRepository.Get(c => c.Name.Contains(arg));
+            if (companys.Any())
+                foreach (var company in companys)
```
please add {}. It's hard to read

stas-i commented on MVCTask/MVCTask.Core/Services/SearchService.cs:
```diff
@@ -0,0 +1,55 @@
+            if (companys.Any())
+                foreach (var company in companys)
+                    result.AddRange(_userRepository.Get(u => u.CompanyId == company.Id));
+            var users = _userRepository.Get(u => u.Name.Contains(arg) | u.Surname.Contains(arg) |
```
Can we replace it with one query?

stas-i commented on MVCTask/MVCTask.Core/Services/TitleService.cs:
```diff
@@ -0,0 +1,44 @@
+
+        public void RemoveTitle(int userId, string title)
+        {
+            _titleRepository.Remove(_titleRepository.Get(t => (t.UserId == userId) & (t.Name == title)).First());
```
Please split in two lines

stas-i commented on MVCTask/MVCTask.Core/Services/TitleService.cs:
```diff
@@ -0,0 +1,44 @@
+                               throw new ArgumentNullException(nameof(titleRepository),
+                                   $"{nameof(titleRepository)} cannot be null.");
+        }
+
```
Empty line

stas-i commented on MVCTask/MVCTask.Core/Services/UserService.cs:
```diff
@@ -0,0 +1,72 @@
+
+        public Task<string> GetFileUrlByIdAsync(int userId)
+        {
+            return _userRepository.AsyncGetFileUrl(userId);
```
GetFileUriAsync

stas-i commented on MVCTask/MVCTask.Core/Services/ImageService.cs:
```diff
@@ -0,0 +1,34 @@
+{
+    public class ImageService : IImageService
+    {
+        public async Task<byte[]> GetImageAsync(string path)
```
Please validate path before processing

stas-i commented on MVCTask/MVCTask.Data/Interface/IUserRepository.cs:
```diff
@@ -0,0 +1,11 @@
+    public interface IUserRepository : IRepository<User>
+    {
+        Task<string> AsyncGetFileUrl(int userId);
+        Task<User> AsyncFindUser(int userId);
```
Not used

stas-i commented on MVCTask/MVCTask.Data/Model/Company.cs:
```diff
@@ -0,0 +1,8 @@
+    public class Company : Entity
+    {
+        public string Name { get; set; }
+        public int MaxCounOfUsers { get; set; }
```
CounT - typo

stas-i commented on MVCTask/MVCTask.Data/Repositores/BaseRepository.cs:
```diff
@@ -0,0 +1,73 @@
+
+        protected BaseRepository(DbContext context)
+        {
+            _context = context ?? throw new ArgumentNullException("context");
```
nameof

stas-i commented on MVCTask/MVCTask.Data/Repositores/BaseRepository.cs:
```diff
@@ -0,0 +1,73 @@
+
+        public void Remove(TEntity item)
+        {
+            //_dbSet.Remove(item);
```
commented code

stas-i commented on MVCTask/MVCTask.Data/Repositores/CompanyRepository.cs:
```diff
@@ -0,0 +1,28 @@
+                var result = Get(c => c.Id == companyId).First().Name;
+                return result;
+            }
+            catch (Exception e)
```
Exception driven development - re implement without try/catch

stas-i commented on MVCTask/MVCTask.Data/Repositores/TitleRepository.cs:
```diff
@@ -0,0 +1,28 @@
+        public IEnumerable<string> GetNamesByUserId(int userId)
+        {
+            try
+            {
```
The same as in company repository

stas-i commented on MVCTask/MVCTask.Data/Repositores/UserRepository.cs:
```diff
@@ -0,0 +1,25 @@
+
+        public Task<User> AsyncFindUser(int userId)
+        {
+            return Task.Run(() => FindById(userId));
```
Unacceptable))
https://msdn.microsoft.com/en-us/library/jj819165(v=vs.113).aspx

stas-i commented on MVCTask/MVCTask/App_Start/StructuremapMvc.cs:
```diff
@@ -0,0 +1,62 @@
+	    public static void End() {
+            StructureMapDependencyScope.Dispose();
+        }
+		
```
tabs still there

stas-i commented on MVCTask/MVCTask/Content/SiteStyle.css:
```diff
@@ -0,0 +1,27 @@
+﻿.list-center-unmarked { }
```
Can SiteStyle.css and Site.css  live in one file?

stas-i commented on MVCTask/MVCTask/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,220 @@
+            _searchService = searchService;
+            _userService = userService;
+            _titelsServise = titelsServise;
+            _maxCount = int.Parse(ConfigurationManager.AppSettings["MaxCountOfUsersOnPage"]);
```
Please create interface for configuration (e.g. IConfig) and implementation (e.g. IConfigurationManagerCofig). What if tomorrow we will need to read this setting from file or environment or some api?

stas-i commented on MVCTask/MVCTask/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,220 @@
+
+            return View(new UsersViewModel
+            {
+                UserModels = (PagedList<UserModel>) users.ToPagedList(page.GetValueOrDefault(1), _maxCount)
```
Please split it in to several rows. Since You are going to add pagination find and index should be in one method and You should split users in to pages on db level

stas-i commented on MVCTask/MVCTask/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,220 @@
+        {
+            var model = new NewEditUserModel
+            {
+                Surname = "",
```
Do we need this initialization

stas-i commented on MVCTask/MVCTask/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,220 @@
+            ViewBag.Title = "New user";
+            if (id > 0)
+            {
+                var userToEdit = _userService.FindUserById(id.Value);
```
Use include

stas-i commented on MVCTask/MVCTask/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,220 @@
+                Email = "",
+                Title = new List<string> {""}
+            };
+            ViewBag.Title = "New user";
```
Move to else

stas-i commented on MVCTask/MVCTask/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,220 @@
+        }
+
+        [HttpPost]
+        public ActionResult CreateUser(HttpPostedFileBase file, NewEditUserModel model, FormCollection formCollection)
```
Remove FormCollection

stas-i commented on MVCTask/MVCTask/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,220 @@
+        {
+            if (ModelState.IsValid)
+            {
+                var path = "";
```
string path;

stas-i commented on MVCTask/MVCTask/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,220 @@
+                }
+                else if (file != null)
+                {
+                    var fil = Path.GetFileName(file.FileName);
```
What is fil

stas-i commented on MVCTask/MVCTask/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,220 @@
+                else if (file != null)
+                {
+                    var fil = Path.GetFileName(file.FileName);
+                    path = Server.MapPath(ConfigurationManager.AppSettings["PhotoPath"] + fil);
```
Move all settings in Your own config. It's hard to manage all settings when You are getting them directly from ConfigurationManager

stas-i replied:
Path.Combine

stas-i commented on MVCTask/MVCTask/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,220 @@
+                    _userService.UpdateUser(model.Id, model.Name, model.Surname, model.Email,
+                        model.BirthDate.GetValueOrDefault(),
+                        model.CompanyId, path);
+                    var oldCountOfTitels = _titelsServise.GetTitelsByUserId(model.Id);
```
It's not count

stas-i commented on MVCTask/MVCTask/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,220 @@
+                        model.BirthDate.GetValueOrDefault(),
+                        model.CompanyId, path);
+                    var oldCountOfTitels = _titelsServise.GetTitelsByUserId(model.Id);
+                    foreach (var oldCountOfTitel in oldCountOfTitels)
```
It's business logic

stas-i commented on MVCTask/MVCTask/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,220 @@
+                    var oldCountOfTitels = _titelsServise.GetTitelsByUserId(model.Id);
+                    foreach (var oldCountOfTitel in oldCountOfTitels)
+                        _titelsServise.RemoveTitle(model.Id, oldCountOfTitel);
+                    foreach (var s in model.Title)
```
Always write { } even if they could be omitted - it's hard to read

stas-i commented on MVCTask/MVCTask/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,220 @@
+                        _titelsServise.RemoveTitle(model.Id, oldCountOfTitel);
+                    foreach (var s in model.Title)
+                        if (s.Length > 0)
+                            _titelsServise.CreateTitle(s, model.Id);
```
It's not the best solution)) You should not delete unmodified 'тители'

stas-i commented on MVCTask/MVCTask/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,220 @@
+                else
+                {
+                    _userService.CreateUser(model.Name, model.Surname, model.Email, model.BirthDate.GetValueOrDefault(),
+                        model.CompanyId, path);
```
Best practice is to return id from Insert method

stas-i commented on MVCTask/MVCTask/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,220 @@
+                return RedirectToAction("Index", "Users");
+            }
+
+            model.Title = new List<string> {""};
```
Why????

stas-i commented on MVCTask/MVCTask/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,220 @@
+        [HttpDelete]
+        public ActionResult DeleteUser(int id)
+        {
+            _userService.DeleteUser(id);
```
What if user alredy does not exist?

stas-i commented on MVCTask/MVCTask/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,220 @@
+
+        private List<UserModel> GetAllUsers()
+        {
+            var someUser = _userService.GetUsers();
```
bad variables names)

stas-i commented on MVCTask/MVCTask/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,220 @@
+            };
+        }
+
+        private List<UserModel> GetAllUsers()
```
Can You launch SQL Profiler and open your index page? And then fix it please

stas-i commented on MVCTask/MVCTask/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,220 @@
+            {
+                CompanyName = company.Name,
+                Id = company.Id,
+                MaxCountOfUsers = company.MaxCounOfUsers
```
Why do we need two max counts? (In settings and in model)

stas-i commented on MVCTask/MVCTask/Filters/LogActionFilter.cs:
```diff
@@ -0,0 +1,48 @@
+                return;
+            }
+
+            filterContext.HttpContext.Session["log"] = string.Format("{0} Action {1} starts",
```
Looks like copy paste - could we extract some log writer

stas-i commented on MVCTask/MVCTask/Scripts/NewEditUserScripts.js:
```diff
@@ -0,0 +1,18 @@
+};
+
+function RemoveTitelInput(id) {
+    console.log(id);
```
please remove console.log

stas-i commented on MVCTask/MVCTask/Scripts/NewEditUserScripts.js:
```diff
@@ -0,0 +1,18 @@
+﻿function AddNewTitleEdit() {
```
please use IIFE

stas-i commented on MVCTask/MVCTask/Validator/UserValidator.cs:
```diff
@@ -0,0 +1,52 @@
+            RuleFor(u => u.Surname).NotEmpty().MaximumLength(50);
+            RuleFor(u => u.Email).NotEmpty().EmailAddress().Must((user, email) =>
+            {
+                var emails = _userRepository.Get().Select(u => u.Email).ToList();
```
Please extract private method. And email should be checked on db level

stas-i commented on MVCTask/MVCTask/Validator/UserValidator.cs:
```diff
@@ -0,0 +1,52 @@
+                    if (user.Id > 0)
+                        if (user.CompanyId == companyId)
+                            return true;
+                    return _userRepository.Get(u => u.CompanyId == companyId).Count() <
```
Looks like we should compare in with setting from config

stas-i commented on MVCTask/MVCTask/Views/Shared/Log.cshtml:
```diff
@@ -0,0 +1,10 @@
+    {
+        HttpContext.Current.Session.Add("log", "");
+    }
+    var log = HttpContext.Current.Session["log"].ToString();
```
Can we store list of strings under Session["log"] ?

stas-i commented on MVCTask/MVCTask/Views/Users/Index.cshtml:
```diff
@@ -0,0 +1,125 @@
+    <link href="~/Content/SiteStyle.css" rel="stylesheet"/>
+    <script src="~/Scripts/IndexScripts.js"></script>
+    <link href="~/Content/bootstrap.min.css" rel="stylesheet"/>
+    <link href="~/Content/PagedList.css" rel="stylesheet"/>
```
Since we have bundles - can we use them?

stas-i commented on MVCTask/MVCTask/Views/Users/Index.cshtml:
```diff
@@ -0,0 +1,125 @@
+</head>
+<body>
+<div>
+    @using (Html.BeginForm("Find", "Users"))
```
Why do we need to draw list of users inside form?

stas-i commented on MVCTask/MVCTask/Views/Users/Index.cshtml:
```diff
@@ -0,0 +1,125 @@
+
+            </li>
+            <li>
+                <input type="button" value="Find" onclick="FindUsers();"/>
```
Can we submit form and send get request without js?

stas-i reviewed (commented):
Optimize Your BL/DAL reduse number of queries to db

JAKLI78 committed:
rename some methods and vars
now send model in create/update user methods
fix Exceptions when load image
remove unused method from userRepository interface
remove try/catch in companyRepository and titleRepository
remove tabs from StructureMapMvc file
move site styles in one file
add custom configuration manager
remove unused formCollection from usersController
now db contain relative links to the photo
add additional config file
reduce count of connections to db when loads users list page
max count of users in company now only in config file
email validation now on db level

JAKLI78 committed:
add log service

JAKLI78 committed:
add configuration in container

JAKLI78 committed:
now log in list of strings

JAKLI78 committed:
remove 'titel'

JAKLI78 committed:
code improves
move config dependency to ctor insted of property

JAKLI78 committed:
code improves
move config dependency to ctor insted of property
